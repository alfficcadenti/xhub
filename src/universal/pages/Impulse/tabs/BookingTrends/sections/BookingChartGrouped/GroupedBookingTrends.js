import React, {useState} from 'react';
import {
    CartesianGrid,
    XAxis,
    YAxis,
    Tooltip,
    Legend,
    ResponsiveContainer,
    ComposedChart, ReferenceArea, Area, ReferenceLine, Line
} from 'recharts';
import moment from 'moment';

import {BRANDS_CHART, CHART_COLORS, LOBS_CHART} from '../../../../constants';
import {startTime, endTime, getColor, getDefaultTimeInterval, getTimeIntervals, isValidTimeInterval} from '../../../../impulseHandler';
import AnomalyLabel from '../BookingChart/AnomalyLabel';
import ReferenceLabel from '../../../../../../components/ReferenceLabel';
import '/src/universal/pages/Impulse/tabs/BookingTrends/sections/BookingChart/styles.less';

const GroupedBookingTrends = ({
    data = [],
    setStartDateTime,
    setEndDateTime,
    setChartSliced,
    daysDifference,
    setDaysDifference, annotations,
    setTableData,
    anomalies,
    setAnomalyTableData,
    timeInterval,
    setTimeInterval, setTimeIntervalOpts,
    activeIndex,
    setIsResetClicked,
    setIsChartSliceClicked
}) => {
    const [left, setLeft] = useState('dataMin');
    const [right, setRight] = useState('dataMax');
    const [refAreaLeft, setRefAreaLeft] = useState('');
    const [refAreaRight, setRefAreaRight] = useState('');

    const [hiddenKeys, setHiddenKeys] = useState([]);
    const [highlightedKey, setHighlightedKey] = useState('');

    const POS_CHART = data.length > 0 ? Object.getOwnPropertyNames(data[0]).slice(1) : [];

    const formatDateTimeLocal = (date) => moment(date).format('MM/DD HH:mm');

    const CustomTooltip = ({active, payload}) => {
        const TIMEZONE = moment().tz(moment.tz.guess()).format('z');
        if (active && payload?.[0]?.payload) {
            return (<div className="custom-tooltip">
                <span className="label">{`${formatDateTimeLocal(payload[0].payload.time)} ${TIMEZONE}`}</span>
                {payload.map((item) => (
                    <div>
                        <span className="label" style={{color: item.dataKey === highlightedKey ? item.stroke : ''}}>{`${item.dataKey} : ${item.value}`} </span>
                    </div>
                ))}
            </div>);
        }
        return null;
    };

    const handleOnMouseEnter = (e) => {
        const {dataKey} = e;
        setHighlightedKey(dataKey);
    };

    const handleOnMouseLeave = () => {
        setHighlightedKey('');
    };

    const renderChart = ({name, color}, idx) => {
        const fill = `url(#${idx})`;
        return data?.[0]?.hasOwnProperty(name) ? <Area type="monotone" dataKey={name} yAxisId={1} stroke={color} fillOpacity={1} fill={fill} hide = {hiddenKeys.includes(name)} onMouseEnter = {handleOnMouseEnter} onMouseLeave = {handleOnMouseLeave}/>
            : '';
    };

    const renderLineChart = (name, idx) => {
        return data?.[0]?.hasOwnProperty(name) ? <Line type="monotone" dataKey={name} stroke={CHART_COLORS[idx]} yAxisId={1} strokeWidth={1.5} dot={false} hide = {hiddenKeys.includes(name)} onMouseEnter = {handleOnMouseEnter} onMouseLeave = {handleOnMouseLeave}/>
            : '';
    };

    const resetGraphToDefault = () => {
        setStartDateTime(startTime());
        setEndDateTime(endTime());
        setChartSliced(false);
        setDaysDifference(3);
        setIsResetClicked(true);
        if (!isValidTimeInterval(startTime(), endTime(), timeInterval)) {
            setTimeInterval('5m');
            setTimeIntervalOpts(['15m', '30m', '1h']);
        } else {
            setTimeIntervalOpts(getTimeIntervals(startTime(), endTime(), timeInterval));
        }
    };

    const zoomIn = () => {
        if (refAreaLeft === refAreaRight || refAreaRight === '') {
            setRefAreaLeft('');
            setRefAreaRight('');
            return;
        }

        let nextRefAreaLeft = refAreaLeft;
        let nextRefAreaRight = refAreaRight;

        if (moment(refAreaLeft).isAfter(refAreaRight)) {
            [nextRefAreaLeft, nextRefAreaRight] = [refAreaRight, refAreaLeft];
        }
        const leftMomentUtc = moment(nextRefAreaLeft).utc();
        const rightMomentUtc = moment(nextRefAreaRight).utc();
        setRefAreaLeft('');
        setRefAreaRight('');
        setLeft(nextRefAreaLeft);
        setRight(nextRefAreaRight);
        setStartDateTime(leftMomentUtc);
        setEndDateTime(rightMomentUtc);
        setChartSliced(true);
        setIsChartSliceClicked(true);
        setDaysDifference(rightMomentUtc.diff(leftMomentUtc, 'days'));
        setTableData([]);
        if (!isValidTimeInterval(leftMomentUtc, rightMomentUtc, timeInterval)) {
            const newInterval = getDefaultTimeInterval(leftMomentUtc, rightMomentUtc);
            setTimeInterval(newInterval);
            setTimeIntervalOpts(getTimeIntervals(leftMomentUtc, rightMomentUtc, newInterval));
        } else {
            setTimeIntervalOpts(getTimeIntervals(leftMomentUtc, rightMomentUtc, timeInterval));
        }
    };

    const handleLegendClick = (e) => {
        if (e && e.dataKey) {
            const nextHiddenKeys = [...hiddenKeys];
            const foundIdx = hiddenKeys.findIndex((h) => h === e.dataKey);
            if (foundIdx > -1) {
                nextHiddenKeys.splice(foundIdx, 1);
            } else {
                nextHiddenKeys.push(e.dataKey);
            }
            setHiddenKeys(nextHiddenKeys);
        }
    };

    const getGradient = ({color}, idx) => (
        <linearGradient id={idx} key={idx} x1="0" y1="0" x2="0" y2="1">
            <stop offset="2%" stopColor={color} stopOpacity={0.5}/>
            <stop offset="98%" stopColor={color} stopOpacity={0}/>
        </linearGradient>
    );

    const renderGradient = () => {
        switch (activeIndex) {
            case 1:
                return (
                    <defs>
                        {BRANDS_CHART.map(getGradient)}
                    </defs>
                );
            case 2:
                return (
                    <defs>
                        {LOBS_CHART.map(getGradient)}
                    </defs>
                );
            default:
                return ('');
        }
    };

    const renderMultiViewPanel = () => {
        switch (activeIndex) {
            case 1:
                return (
                    BRANDS_CHART.map(renderChart)
                );
            case 2:
                return (
                    LOBS_CHART.map(renderChart)
                );
            case 3:
                return (
                    POS_CHART.map(renderLineChart)
                );
            default:
                return ('');
        }
    };

    return (
        <div className="bookings-container-box">
            <div className="reset-div"
                title={daysDifference === 3
                    ? 'Click to reset graph to default 3 day date time range (Disabled as Default Range Selected) '
                    : 'Click to reset graph to default 3 day date time range'}
            >
                <button
                    type="button"
                    className={'btn btn-default reset-btn'}
                    disabled={daysDifference === 3}
                    onClick={resetGraphToDefault}
                >
                    {'Reset Graph'}
                </button>
            </div>
            <ResponsiveContainer width="100%" height="70%">
                <ComposedChart
                    data={data}
                    margin={{top: 10, right: 30, left: 0, bottom: 0}}
                    onMouseDown={(e) => e && setRefAreaLeft(e.activeLabel)}
                    onMouseMove={(e) => refAreaLeft && setRefAreaRight(e.activeLabel)}
                    onMouseUp={zoomIn}
                    cursor="crosshair"
                >
                    {renderGradient()}
                    {renderMultiViewPanel()}
                    <XAxis allowDataOverflow type="number" scale="time" dataKey="time" tick={{fontSize: 10}} tickFormatter={formatDateTimeLocal} domain={[left, right]}/>
                    <YAxis allowDataOverflow yAxisId={1} tick={{fontSize: 10}} type="number"/>
                    <CartesianGrid strokeDasharray="3 3"/>
                    <Tooltip offset={15} content={CustomTooltip}/>
                    {
                        anomalies && anomalies.map((anomaly) => (
                            <ReferenceLine
                                key={Math.random()}
                                yAxisId={1}
                                x={anomaly.time}
                                label={<AnomalyLabel anomaly={anomaly} setAnomalyTableData={setAnomalyTableData} category={anomaly.category}/>}
                                stroke={getColor(anomaly)}
                                strokeDasharray="3 3"
                            />
                        ))
                    }
                    {
                        annotations && annotations.map((annotation) => (
                            <ReferenceLine
                                key={Math.random()}
                                yAxisId={1}
                                x={annotation.incidentTime}
                                label={<ReferenceLabel annotation={annotation} isImpulse setTableData={setTableData}/>}
                                stroke={'red'}
                                strokeDasharray="3 3"
                                isFront
                            />
                        ))
                    }
                    {
                        (refAreaLeft && refAreaRight)
                            ? <ReferenceArea yAxisId={1} x1={refAreaLeft} x2={refAreaRight} strokeOpacity={0.3}/>
                            : null
                    }
                    <Legend onClick={handleLegendClick}/>
                </ComposedChart>
            </ResponsiveContainer>
        </div>
    );
};

export default React.memo(GroupedBookingTrends);
