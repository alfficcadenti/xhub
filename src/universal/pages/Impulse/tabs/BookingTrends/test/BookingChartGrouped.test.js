import GroupedBookingTrends from '../sections/BookingChartGrouped/GroupedBookingTrends';
import {render, shallow} from 'enzyme';
import React from 'react';
import chai, {expect} from 'chai';
import chaiJestSnapshot from 'chai-jest-snapshot';
import fakeData from './dataByBrand.test.json';
chai.use(chaiJestSnapshot);

describe('Booking Trend component testing', () => {
    let wrapper;

    beforeEach(() => {
        chaiJestSnapshot.setFilename(`${__filename}.snap`);
        wrapper = shallow(
            <GroupedBookingTrends data ={fakeData} />
        );
    });
    afterEach(() => {
        wrapper.unmount();
    });


    it('renders successfully', () => {
        expect(wrapper).to.have.length(1);
    });

    it('matches the snapshot', () => {
        chaiJestSnapshot.setTestName('matches the snapshot');
        const wrapper1 = render(<GroupedBookingTrends data={fakeData}/>);
        expect(wrapper1).to.matchSnapshot();
    });
});
