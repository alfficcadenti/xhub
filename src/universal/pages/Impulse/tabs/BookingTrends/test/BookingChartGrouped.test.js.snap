// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
initialize {
  "0": Node {
    "attribs": Object {
      "class": "bookings-container-box",
    },
    "children": Array [
      Node {
        "attribs": Object {
          "class": "download-reset-container",
          "data-html2canvas-ignore": "true",
        },
        "children": Array [
          Node {
            "attribs": Object {},
            "children": Array [
              Node {
                "data": "Download graph",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "button",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Node {
              "attribs": Object {
                "class": "btn btn-default reset-btn",
                "title": "Click to reset graph to default 3 day date time range",
                "type": "button",
              },
              "children": Array [
                Node {
                  "data": "Reset Graph",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "button",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
                "title": undefined,
                "type": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
                "title": undefined,
                "type": undefined,
              },
            },
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {},
            "x-attribsPrefix": Object {},
          },
          Node {
            "attribs": Object {
              "class": "btn btn-default reset-btn",
              "title": "Click to reset graph to default 3 day date time range",
              "type": "button",
            },
            "children": Array [
              Node {
                "data": "Reset Graph",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "button",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": Node {
              "attribs": Object {},
              "children": Array [
                Node {
                  "data": "Download graph",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "button",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
              "title": undefined,
              "type": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
              "title": undefined,
              "type": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Node {
          "attribs": Object {
            "class": "recharts-responsive-container",
            "style": "width:100%;height:70%",
          },
          "children": Array [
            Node {
              "attribs": Object {
                "style": "position:absolute;width:0;height:0;visibility:hidden;display:none",
              },
              "children": Array [],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "style": undefined,
              },
              "x-attribsPrefix": Object {
                "style": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "style": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "data-html2canvas-ignore": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "data-html2canvas-ignore": undefined,
        },
      },
      Node {
        "attribs": Object {
          "class": "recharts-responsive-container",
          "style": "width:100%;height:70%",
        },
        "children": Array [
          Node {
            "attribs": Object {
              "style": "position:absolute;width:0;height:0;visibility:hidden;display:none",
            },
            "children": Array [],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "style": undefined,
            },
            "x-attribsPrefix": Object {
              "style": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Node {
          "attribs": Object {
            "class": "download-reset-container",
            "data-html2canvas-ignore": "true",
          },
          "children": Array [
            Node {
              "attribs": Object {},
              "children": Array [
                Node {
                  "data": "Download graph",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "button",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Node {
                "attribs": Object {
                  "class": "btn btn-default reset-btn",
                  "title": "Click to reset graph to default 3 day date time range",
                  "type": "button",
                },
                "children": Array [
                  Node {
                    "data": "Reset Graph",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "button",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                  "title": undefined,
                  "type": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                  "title": undefined,
                  "type": undefined,
                },
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            Node {
              "attribs": Object {
                "class": "btn btn-default reset-btn",
                "title": "Click to reset graph to default 3 day date time range",
                "type": "button",
              },
              "children": Array [
                Node {
                  "data": "Reset Graph",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "button",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": Node {
                "attribs": Object {},
                "children": Array [
                  Node {
                    "data": "Download graph",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "button",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
                "title": undefined,
                "type": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
                "title": undefined,
                "type": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "data-html2canvas-ignore": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "data-html2canvas-ignore": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "style": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
